### PR TITLE
Add index check for Transpose's type inference function

### PR DESCRIPTION
**Pull request by snnn**
@gramalingam Before passing an index to protobuf's repeated field, you must check if the index is valid first.  Otherwise, it's a security issue.

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -474,16 +477,37 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-                 --i) {
+          bool has_perm_attr = getRepeatedAttribute(ctx, "perm", perm);
+          if (!has_perm_attr) {
+            for (int i = shape.dim_size() - 1; i >= 0; --i) {
               perm.push_back(i);
             }
           }
+          else if(!perm.empty()){
+              for (int64_t fromDimIndex : perm)
+                  if (shape.dim_size() <= fromDimIndex || fromDimIndex >= std::numeric_limits<int>::max()) {
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: if (! ((0 <= fromDimIndex) && (fromDimIndex < shape.dim_size()))

Comment (snnn) [reply onnx/defs/tensor/defs.cc]: if (fromDimIndex < 0 || shape.dim_size() <= fromDimIndex || fromDimIndex >= std::numeric_limits<int>::max())  ?

Comment (snnn) [reply onnx/defs/tensor/defs.cc]: Thanks. Will change it.

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -296,8 +296,11 @@ ONNX_OPERATOR_SET_SCHEMA(
             if (!ctx.getInputType(0)->tensor_type().has_shape()) {
               return;
             }
-            const auto& splitDim =
-                ctx.getInputType(0)->tensor_type().shape().dim(axis);
+            const auto& shape = ctx.getInputType(0)->tensor_type().shape();
+            if (axis >= shape.dim_size()) {
+                return;
```
Comment (bddppq) [inline onnx/defs/tensor/defs.cc]: you should throw error here instead of silently return

File: onnx/defs/tensor/defs.cc [hunk 3]
```diff
@@ -474,16 +477,38 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-                     ctx.getInputType(0)->tensor_type().shape().dim_size() - 1;
-                 i >= 0;
-                 --i) {
+          bool has_perm_attr = getRepeatedAttribute(ctx, "perm", perm);
+          if (!has_perm_attr) {
+            for (int i = shape.dim_size() - 1; i >= 0; --i) {
               perm.push_back(i);
             }
           }
+          else if(!perm.empty()){
```
Comment (bddppq) [inline onnx/defs/tensor/defs.cc]: please run clang-format

File: onnx/defs/tensor/defs.cc [hunk 4]
```diff
@@ -474,16 +477,38 @@ ONNX_OPERATOR_SET_SCHEMA(
           if (!hasNInputShapes(ctx, 1)) {
             return;
           }
-
+          auto input_type = ctx.getInputType(0);
+          if (!input_type->has_tensor_type()) {
```
Comment (bddppq) [inline onnx/defs/tensor/defs.cc]: `hasNInputShapes` already checks `has_tensor_type`

File: onnx/defs/tensor/defs.cc [hunk 5]
```diff
@@ -474,16 +477,38 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+              throw InferenceError("Input 0 expected to have tensor type");
```
Comment (bddppq) [inline onnx/defs/tensor/defs.cc]: use the macro `fail_shape_inference`

File: onnx/defs/tensor/defs.cc [hunk 6]
```diff
@@ -474,16 +477,38 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+                      std::ostringstream oss;
+                      oss << "Invalid attribute perm {" << perm[0];
+                      for (size_t i=1;i!=perm.size();++i) {
+                          oss << ", " << perm[i];
+                      }
+                      oss << "}, input shape = {";
+                      if (shape.dim_size() > 0) {
+                          oss<< shape.dim(0).dim_value();
+                          for (int i = 1; i != shape.dim_size(); ++i) {
+                              oss << ", " << shape.dim(i).dim_value();
```
Comment (bddppq) [inline onnx/defs/tensor/defs.cc]: Why do you need to print the dim values? what user should know is `shape.dim_size()` right?

File: onnx/defs/tensor/defs.cc (hunk 2 above)
Comment (snnn) [reply onnx/defs/tensor/defs.cc]: Fixed.

File: onnx/defs/tensor/defs.cc (hunk 3 above)
Comment (snnn) [reply onnx/defs/tensor/defs.cc]: Fixed.

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (snnn) [reply onnx/defs/tensor/defs.cc]: Fixed.

File: onnx/defs/tensor/defs.cc (hunk 5 above)
Comment (snnn) [reply onnx/defs/tensor/defs.cc]: Fixed.

File: onnx/defs/tensor/defs.cc (hunk 6 above)
Comment (snnn) [reply onnx/defs/tensor/defs.cc]: It's not so obvious. Because they may be inferred from another node.  Currently, there is no way to view it.

**Comment by gramalingam** [comment]
Hmmm … the Travis CI build keeps timing out … any idea what we can do?

**Pull request merged**